### PR TITLE
Mtt aaa

### PR DESCRIPTION
**Pull request by renukavinay**
Added dataclass, api and unittests for AAA, Radius and Tacacs in vmanage 20.12 release.

### CODE REVIEW

File: vmngclient/api/mtt_aaa_api.py [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+            "accounting": false,
+            "radiusServers": "server1"
+
+        returns bool depending on the api post call
+        """
+        url_path = "/dataservice/admin/aaa"
+        data = asdict(tenant_aaa)  # type: ignore
+
+        response = self.session.post(url=url_path, json=data)
+        logger.info(response)
```
Comment (sbasan) [inline vmngclient/api/mtt_aaa_api.py:41]: Not sure if this is needed. Will log something like "<Response [200]>"
vManageSession logs each response (with details) on debug level.

File: vmngclient/api/mtt_aaa_api.py [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+
+    def get_aaa(self) -> TenantAAA:
+        """
+        Returns the Tenant AAA
+        :param aaa:
+        :return:
+        """
+        url_path = "/dataservice/admin/aaa"
+        tenant_aaa = self.session.get_data(url_path)
+        logger.debug(f"Tenant AAA: {tenant_aaa}")
```
Comment (sbasan) [inline vmngclient/api/mtt_aaa_api.py:52]: Same as for response logging, Whole response including parsed json data is logged by vManageSession.

File: vmngclient/api/mtt_aaa_api.py [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+
+    def __str__(self) -> str:
+        return str(self.session)
+
+    def aaa_exists(self) -> bool:
+        url_path = "/dataservice/admin/aaa"
+        if not self.session.get_data(url_path):
+            return False
+        else:
+            return True
```
Comment (igosoft) [inline vmngclient/api/mtt_aaa_api.py:30]: it's good place for one-liner

File: vmngclient/api/mtt_aaa_api.py [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+            "auditDisable": false,
+            "accounting": false,
+            "radiusServers": "server1"
+
+        returns bool depending on the api post call
+        """
+        url_path = "/dataservice/admin/aaa"
+        data = asdict(tenant_aaa)  # type: ignore
+        response = self.session.post(url=url_path, json=data)
+        return True if response.status_code == 200 else False
```
Comment (igosoft) [inline vmngclient/api/mtt_aaa_api.py:47]:
```suggestion
        return True if response.status_code == HTTPStatus.OK else False
```
from http import HTTPStatus

File: vmngclient/api/mtt_aaa_api.py [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+
+    def get_aaa(self) -> TenantAAA:
+        """
+        Returns the Tenant AAA
+        :param aaa:
+        :return:
+        """
+        url_path = "/dataservice/admin/aaa"
+        tenant_aaa = self.session.get_data(url_path)
+        logger.debug(f"Tenant AAA: {tenant_aaa}")
```
Comment (igosoft) [inline vmngclient/api/mtt_aaa_api.py:57]: Do we need to put logs getters ?

File: vmngclient/api/mtt_aaa_api.py [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+    def del_aaa(self) -> bool:
+        """
+        Delete aaa works only for tenants
+        :return:
+        """
+        url_path = "/dataservice/admin/aaa"
+        logger.debug(f"Deleting AAA on {self.session.get_tenant_id()}.")
+        response = self.session.delete(url_path)
+        if not self.aaa_exists():
+            raise AAAConfigNotPresent("No Aaa config present for Tenant")
```
Comment (igosoft) [inline vmngclient/api/mtt_aaa_api.py:69]:
```suggestion
            raise AAAConfigNotPresent(f"No AAA config present for Tenant id={id}")
```

File: vmngclient/api/mtt_aaa_api.py [hunk 7]
```diff
@@ -0,0 +1,186 @@
...
+        logger.debug(f"Tenant AAA: {tenant_aaa}")
+        return create_dataclass(TenantAAA, tenant_aaa)
+
+    def del_aaa(self) -> bool:
+        """
+        Delete aaa works only for tenants
+        :return:
+        """
+        url_path = "/dataservice/admin/aaa"
+        logger.debug(f"Deleting AAA on {self.session.get_tenant_id()}.")
```
Comment (igosoft) [inline vmngclient/api/mtt_aaa_api.py:66]:
```suggestion
        id = self.session.get_tenant_id()
        logger.debug(f"Deleting AAA on {id}.")
```

File: vmngclient/api/mtt_aaa_api.py [hunk 8]
```diff
@@ -0,0 +1,186 @@
...
+    Used to configure  mtt tenant remote aaa radius servers
+    """
+
+    def __init__(self, session: vManageSession) -> None:
+        self.session = session
+
+    def __str__(self) -> str:
+        return str(self.session)
+
+    def add_radius(self, radius_server: TenantRadiusServer) -> bool:
```
Comment (igosoft) [inline vmngclient/api/mtt_aaa_api.py:94]:
```suggestion
    def add(self, radius_server: TenantRadiusServer) -> bool:
```

File: vmngclient/api/mtt_aaa_api.py [hunk 9]
```diff
@@ -0,0 +1,186 @@
...
+        Create RADIUS for tenant
+        :param radius_server:
+        :return:
+        """
+        url_path = "/dataservice/admin/radius"
+        data = asdict(radius_server)  # type: ignore
+        response = self.session.post(url=url_path, json=data)
+        return True if response.status_code == 200 else False
+
+    def put_radius(self, radius_server: TenantRadiusServer) -> bool:
```
Comment (igosoft) [inline vmngclient/api/mtt_aaa_api.py:105]:
```suggestion
    def put(self, radius_server: TenantRadiusServer) -> bool:
```

File: vmngclient/api/mtt_aaa_api.py [hunk 10]
```diff
@@ -0,0 +1,186 @@
...
+        edit radius server
+        :param radius_server:
+        :return:
+        """
+        url_path = "/dataservice/admin/radius"
+        data = asdict(radius_server)  # type: ignore
+        response = self.session.put(url=url_path, json=data)
+        return True if response.status_code == 200 else False
+
+    def delete_radius(self) -> bool:
```
Comment (igosoft) [inline vmngclient/api/mtt_aaa_api.py:116]:
```suggestion
    def delete(self) -> bool:
```

File: vmngclient/api/mtt_aaa_api.py [hunk 11]
```diff
@@ -0,0 +1,186 @@
...
+        """
+        edit radius server
+        :param radius_server:
+        :return: True|False
+        """
+        url_path = "/dataservice/admin/radius"
+        response = self.session.delete(url_path)
+        return True if response.status_code == 204 else False
+
+    def get_radius(self) -> TenantRadiusServer:
```
Comment (igosoft) [inline vmngclient/api/mtt_aaa_api.py:126]:
```suggestion
    def get(self) -> TenantRadiusServer:
```

File: vmngclient/api/mtt_aaa_api.py [hunk 12]
```diff
@@ -0,0 +1,186 @@
...
+        response = self.session.post(url=url_path, json=data)
+        return True if response.status_code == 200 else False
+
+    def put_radius(self, radius_server: TenantRadiusServer) -> bool:
+        """
+        edit radius server
+        :param radius_server:
+        :return:
+        """
+        url_path = "/dataservice/admin/radius"
```
Comment (igosoft) [inline vmngclient/api/mtt_aaa_api.py:111]: We can think about moving this line from functions to class member

File: vmngclient/tests/test_mtt_aaa_api.py [hunk 13]
```diff
@@ -0,0 +1,297 @@
...
+    @parameterized.expand([[200, True], [400, False]])
+    @patch("vmngclient.session.Session")
+    @patch("requests.Response")
+    def test_add_aaa(self, status_code, expected_outcome, mock_session, mock_response):
+        # Arrange
+        mock_session.get_data.return_value = self.aaa
+        mock_session.post.return_value = mock_response
+        mock_response.status_code = status_code
+        # Act
+        answer = TenantAaaAPI(mock_session).add_aaa(self.aaa_dataclass)
```
Comment (igosoft) [inline vmngclient/tests/test_mtt_aaa_api.py:54]:
```suggestion
        answer = TenantAaaAPI(mock_session).add(self.aaa_dataclass)
```
simplified method name looks better

File: vmngclient/api/mtt_aaa_api.py [hunk 14]
```diff
@@ -0,0 +1,186 @@
...
+
+    def delete_radius(self) -> bool:
+        """
+        edit radius server
+        :param radius_server:
+        :return: True|False
+        """
+        url_path = "/dataservice/admin/radius"
+        response = self.session.delete(url_path)
+        return True if response.status_code == 204 else False
```
Comment (igosoft) [inline vmngclient/api/mtt_aaa_api.py:124]: 204 is NO_CONTENT https://datatracker.ietf.org/doc/html/rfc7231.html#section-6.3.5
Do we receive 204 as repose status code ?

File: vmngclient/api/mtt_aaa_api.py [hunk 15]
```diff
@@ -0,0 +1,186 @@
...
+    def add_radius(self, radius_server: TenantRadiusServer) -> bool:
+        """
+        Create RADIUS for tenant
+        :param radius_server:
+        :return:
+        """
+        url_path = "/dataservice/admin/radius"
+        data = asdict(radius_server)  # type: ignore
+        response = self.session.post(url=url_path, json=data)
+        return True if response.status_code == 200 else False
```
Comment (igosoft) [inline vmngclient/api/mtt_aaa_api.py:103]: Maybe we should create some wrapper and replace  
"return True if response.status_code == 200 else False"

with response.is_200() or responce.is_ok()

return self.session.post(url=url_path, json=data).is_ok()

File: vmngclient/api/mtt_aaa_api.py (hunk 8 above)
Comment (igosoft) [reply vmngclient/api/mtt_aaa_api.py:94]: public API should be each time simplified if possible

**Review by igosoft — COMMENTED** [review]
Few minor comments

File: vmngclient/api/mtt_aaa_api.py (hunk 14 above)
Comment (renukavinay) [reply vmngclient/api/mtt_aaa_api.py:124]: Yes, the response is 204

File: vmngclient/tests/test_mtt_aaa_api.py (hunk 13 above)
Comment (renukavinay) [reply vmngclient/tests/test_mtt_aaa_api.py:54]: there are three components here: AAA, RADIUS and TACACS. Thats why added respective components.

File: vmngclient/api/mtt_aaa_api.py (hunk 11 above)
Comment (renukavinay) [reply vmngclient/api/mtt_aaa_api.py:124]: there are three components here: AAA, RADIUS and TACACS. Thats why added respective components.

File: vmngclient/api/mtt_aaa_api.py (hunk 8 above)
Comment (renukavinay) [reply vmngclient/api/mtt_aaa_api.py:89]: But this api file is addressing three components AAA, RADIUS and TACACS

File: vmngclient/api/mtt_aaa_api.py (hunk 15 above)
Comment (sbasan) [reply vmngclient/api/mtt_aaa_api.py:97]: Additional method could be confusing.
https://requests.readthedocs.io/en/latest/api/#requests.Response.ok
response.ok - builtin
response.is_ok() - proposed using wrapper

Comment (renukavinay) [reply vmngclient/api/mtt_aaa_api.py]: added wrapper 
def status_ok(func):
    def wrapper(*args, **kwargs):
        response = func(*args, **kwargs)
        return True if response.status_code == 200 else False

    return wrapper

File: vmngclient/api/mtt_aaa_api.py (hunk 4 above)
Comment (renukavinay) [reply vmngclient/api/mtt_aaa_api.py]: Need 204 also for this suite

**Review by JimOverholt — APPROVED** [review]

**Pull request merged**